### PR TITLE
Rename theano.gof objects

### PR DESCRIPTION
**Pull request by brandonwillard**
This PR introduces a few new important refactorings in `theano.gof`.  Most are renamings of basic classes and utility functions that clarify their purpose/use (e.g. `theano.gof.graph.inputs` is renamed to `theano.gof.graph.graph_inputs` so that it's use doesn't clash with ubiquitous local variables also named `inputs`).

More importantly, this PR includes a rename of `Type` to `CType`, which involves some important `CLinker` changes.  Those changes also improve the clarity of `CLinker`'s code by removing a lot of ambiguous logic (e.g. `try` statements) that allow objects to forego an ultimately required C code-generating interface.

### CODE REVIEW

File: theano/compile/ops.py [hunk 1]
```diff
@@ -619,11 +619,11 @@ def numpy_dot(a, b):
     """
     if not isinstance(itypes, (list, tuple)):
         itypes = [itypes]
-    if any(not isinstance(t, theano.Type) for t in itypes):
+    if any(not isinstance(t, theano.CType) for t in itypes):
```
Comment (michaelosthege) [inline theano/compile/ops.py:622]: This reference goes through `theano.*`

File: theano/gof/graph.py [hunk 2]
```diff
@@ -223,27 +223,26 @@ def clone(self):
...
         Parameters
         ----------
-        inputs
-            List of Variable instances to use as inputs.
-        strict : bool
-            If True, the type fields of all the inputs must be equal
-            to the current ones (or compatible, for instance Tensor /
-            GpuArray of the same dtype and broadcastable patterns,
-            in which case they will be converted into current Type), and
+        inputs: list of Variables
```
Comment (michaelosthege) [inline theano/gof/graph.py:230]: In numpydoc, there should be a space to both sides of the colon: https://numpydoc.readthedocs.io/en/latest/format.html#class-docstring

**Review by michaelosthege — CHANGES_REQUESTED** [review]
Looks really good! Just a few details..
This `contextlib.suppress` is a real hidden gem!

File: doc/JaxOps.rst [hunk 3]
```diff
@@ -58,7 +58,7 @@ Here’s an example for the Eye Op.
 .. code:: python
 
    from theano.tensor.basic import Eye
-   @jax_funcify.register(Eye) # The decorater 
+   @jax_funcify.register(Eye) # The decorater
```
Comment (twiecki) [inline doc/JaxOps.rst:61]:
```suggestion
   @jax_funcify.register(Eye) # The decorator
```

**Review by twiecki — APPROVED** [review]

**Comment by twiecki** [comment]
Agree, this looks really great. For your consideration:
Rename `basic` to `core` and `fg` to `function` (so that it's `graph.function.FunctionGraph`. Not super excited about them but tossing them out in case you like them better.

**Review by michaelosthege — APPROVED** [review]
This is great!

I also tend to use "core" submodules, but this codebase uses "basic" in so many places that I don't think we should break with that style.

I noticed a few string formatting problems in the `__author__` fields. But for the sake of reviewability, I think this should be addressed later on.

**Comment by brandonwillard** [comment]
> I also tend to use "core" submodules, but this codebase uses "basic" in so many places that I don't think we should break with that style.

`core` would be better, but, yes, I decided to go with the existing convention.  We can definitely change that though&mdash;although we should do that for all the modules at once.

> I noticed a few string formatting problems in the `__author__` fields. But for the sake of reviewability, I think this should be addressed later on.

We need to completely remove those nonsense dunder fields in the source files.  More specifically, we should move all the `__copyright__`s to the `LICENSE` file&mdash;if they aren't already covered by what's in there.  Otherwise, outside of externally contributed files, authorship is automatically well documented by Git itself, so those `__author__` fields are superfluous.

**Comment by twiecki** [comment]
Let's merge this then. But need to make sure to also fix PyMC3.

**Pull request merged**